desktop-widgets/command_edit.cpp

```diff
@@ -0,0 +1,89 @@
+// SPDX-License-Identifier: GPL-2.0
+
+#include "command_edit.h"
+#include "core/divelist.h"
+
+namespace Command {
+
+template<typename T>
+EditBase<T>::EditBase(const QVector<dive *> &divesIn, T newValue, T oldValue) :
+	value(std::move(newValue)),
+	old(std::move(oldValue)),
+	dives(divesIn.toStdVector())
+{
+	// If there is nothing to do, clear the dives vector.
+	// This signals that no action has to be taken.
+	if (old == value)
+		dives.clear();
+}
+
+// This is quite hackish: we can't use virtual functions in the constructor and
+// therefore can't initialize the list of dives [the values of the dives are
+// accessed by virtual functions]. Therefore, we (mis)use the fact that workToBeDone()
+// is called exactly once before adding the Command to the system and perform this here.
+// To be more explicit about this, we might think about renaming workToBeDone() to init().
+template<typename T>
+bool EditBase<T>::workToBeDone()
+{
+	std::vector<dive *> divesNew;
+	divesNew.reserve(dives.size());
+	for (dive *d: dives) {
+		if (data(d) == old)
+			divesNew.push_back(d);
+	}
+	dives = std::move(divesNew);
+
+	// Create a text for the menu entry. In the case of multiple dives add the number
+	size_t num_dives = dives.size();
+	if (num_dives > 0)
+		//: remove the part in parantheses for %n = 1
+		setText(tr("Edit %1 (%n dive(s))", "", num_dives).arg(fieldName()));
+
+	return num_dives;
+}
+
+template<typename T>
+void EditBase<T>::undo()
+{
+	if (dives.empty())
+		return; // Shouldn't happen
+
+	for (dive *d: dives)
+		set(d, value);
+
+	std::swap(old, value);
+
+	mark_divelist_changed(true);
+}
+
+// The constructors of the base class will be called directly by functions
+// generating the derived classes, owing to the "using" declaration.
+// But they don't have access to the implementation so can't instantiate
+// them. Therefore, explicitly intantiate the constructors here.
```

Comment:
I understand all the words here... I don't think I actually understand what you are telling me.

#InOverMyHead

Reply:
Tried to reword the comment - not sure if the new one is more easily understood. It *is* subtle. Moving the constructor in a header file would probably remove the need of this instantiation, but I'm not sure we want that.